changelog.d/11900.bugfix

```diff
@@ -0,0 +1 @@
+Fix a bug introduced in Synapse 1.37.0rc1 (when knocking was added) where event authorization could fail on valid knocks to a room.
```

Comment:
To ensure I understand, this is causing a user to not be able to go from knock -> leave?

Reply:
argh. Changes to the auth algorithm mean that different implementations reach different conclusions about room state, which can lead to split-braining.

This is exactly the same debate we're currently having elsewhere about stringy PLs: in theory we can't change this without a new room version.

Reply:
I'm highly unconvinced that this particular issue needs spec involvement. 

The path already would have caused different conclusions to be reached, thus we should just fix it and off we go.

Reply:
Note that we do have tests for knocking in Complement. Checking that we can successfully rescind a knock is one of them: https://github.com/matrix-org/complement/blob/154fdc069700bf43694281505da5bf9e2f79bf21/tests/msc2403_test.go#L169-L178

Reply:
@turt2live I'm still slighty confused about which transitions we're fixing here. Note that `RoomVersion.msc2403_knocking` is truthy, so the current condition is equivalent to:

```
membership != Membership.JOIN and membership != Membership.KNOCK
```

I don't think that actually causes a bug here, but please prove me wrong.

Reply:
Let's ignore this thread until the others are resolved please.